File: easy/MatrixDiagonalSum.py
class Solution:
    def diagonalSum(self, mat: List[List[int]]) -> int:
        ans = 0
        cols = len(mat)
        rows = len(mat[0]) - 1
        for i in range(len(mat)):
            if len(mat) % 2 is not 0 and rows - i is i:
                ans += mat[i][rows - i]
            else:
                ans += mat[i][i] + mat[i][rows - i]
        return ans
        #two pointer algo
        #ans = 0
        #end = len(mat) - 1
        #start = 0
        #if len(mat) == 1:
        #    return mat[0][0]
        #while start <= end:
        #    if len(mat) % 2 is not 0 and start == end:
        #        ans += mat[start][start]
        #        return ans
        #    else:
        #        ans += mat[start][start]
        #        ans += mat[start][end]
        #        ans += mat[end][start]
        #        ans += mat[end][end]
        #    start += 1
        #    end -= 1
        #return ans

File: easy/MatrixDiagonalSum.cpp
class Solution {
public:
    int diagonalSum(vector<vector<int>>& mat) {
        int cols = mat.size();
        int ans = 0;
        for(int i = 0; i < cols; i++){
            int len = mat[i].size() - 1;
            //checl if odd num of cols and intersect
            if(cols % 2 != 0 and i == len - i)
                ans += mat[i][len - i];
            else
                ans += mat[i][i] + mat[i][len - i];
        }
        return ans;
    }
};
